commons/src/build_output.rs

```diff
@@ -0,0 +1,709 @@
+use libherokubuildpack::write::{line_mapped, mappers::add_prefix};
+use std::io::Write;
+use std::time::Instant;
+use time::BuildpackDuration;
+
+pub use section::{RunCommand, Section};
+
+/// Build with style
+///
+/// ```rust,no_run
+/// use commons::build_output::{self, RunCommand};
+/// use std::process::Command;
+///
+/// // Announce your buildpack and time it
+/// let timer = build_output::buildpack_name("Buildpack name");
+/// // Do stuff here
+/// timer.done();
+///
+/// // Create section with a topic
+/// let section = build_output::section("Ruby version");
+///
+/// // Output stuff in that section
+/// section.say("Installing");
+/// section.say_with_details("Installing", "important stuff");
+///
+/// // Live stream a progress timer in that section
+/// let mut timer = section.say_with_inline_timer("Installing with progress");
+/// // Do stuff here
+/// timer.done();
+///
+/// // Decorate and format your output
+/// let version = build_output::fmt::value("3.1.2");
+/// section.say(format!("Installing {version}"));
+///
+/// // Run a command in that section with a variety of formatting options
+/// // Stream the output to the user:
+/// section
+///     .run(RunCommand::stream(
+///         Command::new("echo").args(&["hello world"]),
+///     ))
+///     .unwrap();
+///
+/// // Run a command after announcing it. Show a progress timer but don't stream the output :
+/// section
+///     .run(RunCommand::inline_progress(
+///         Command::new("echo").args(&["hello world"]),
+///     ))
+///     .unwrap();
+///
+///
+/// // Run a command with no output:
+/// section
+///     .run(RunCommand::quiet(
+///         Command::new("echo").args(&["hello world"]),
+///     ))
+///     .unwrap();
+///
+/// // Control the display of the command being run:
+/// section
+///     .run(RunCommand::stream(
+///         Command::new("bash").args(&["-c", "exec", "echo \"hello world\""]),
+///     ).with_name("echo \"hello world\""))
+///     .unwrap();
+///```
+
+mod time {
+    use super::{fmt, raw_inline_print};
+    use std::thread::{self, JoinHandle};
+    use std::time::Duration;
+    use std::time::Instant;
+
+    /// Time the entire buildpack execution
+    pub struct BuildpackDuration {
+        pub(crate) start: Instant,
+    }
+
+    impl BuildpackDuration {
+        /// Emit timing details with done block
+        pub fn done_timed(self) {
+            let time = human(&self.start.elapsed());
+            let details = fmt::details(format!("finished in {time}"));
+            println!("- Done {details}");
+        }
+
+        /// Emit done block without timing details
+        #[allow(clippy::unused_self)]
+        pub fn done(self) {
+            println!("- Done");
+        }
+
+        /// Finish without announcing anything
+        #[allow(clippy::unused_self)]
+        pub fn done_silently(self) {}
+    }
+
+    /// Handles outputing inline progress based on timing
+    ///
+    /// i.e.   `- Installing [------] (5.733s)`
```

Comment:
I've seen other tools use ellipsis for this kind of stuff. For example:

`Installing.`
`Installing..`
`Installing...`

Even our own cli does a variation of this. What do you think of that?

Reply:
Pack uses `[=====>     ]` (but then clears them after they're done) here's an example:

![image](https://github.com/heroku/buildpacks-ruby/assets/59744/4c369ce3-84f5-4232-b0df-d41c0796a1a2)

Here's an example using dimmed dots:

![image](https://github.com/heroku/buildpacks-ruby/assets/59744/fee22852-c5cd-47ed-8a4a-e27ea0904d29)


And the raw logs (without colors):

```
$ $(set -pipefail; cd buildpacks/ruby && cargo libcnb package; cd ../..) &&
docker rmi my-image --force  &&
pack build my-image --buildpack target/buildpack/debug/heroku_ruby --path buildpacks/ruby/tests/fixtures/default_ruby
INFO - Reading buildpack metadata...
INFO - Found buildpack heroku/ruby with version 2.0.0.
INFO - Determining automatic cross-compile settings...
INFO - Building binaries (x86_64-unknown-linux-musl)...
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on build directory
   Compiling commons v0.1.0 (/Users/rschneeman/Documents/projects/work/buildpacks/buildpacks-ruby/commons)
   Compiling heroku-ruby-buildpack v0.0.0 (/Users/rschneeman/Documents/projects/work/buildpacks/buildpacks-ruby/buildpacks/ruby)
    Finished dev [unoptimized] target(s) in 17.90s
INFO - Writing buildpack directory...
INFO - Successfully wrote buildpack directory: ../../target/buildpack/debug/heroku_ruby (12.65 MiB)
INFO - Packaging successfully finished!
INFO - Hint: To test your buildpack locally with pack, run: pack build my-image --buildpack ../../target/buildpack/debug/heroku_ruby --path /path/to/application
Untagged: my-image:latest
Deleted: sha256:2e03f8025ffa454017f63cbb64b429efbaa479dcb35f2ea88d98fc8ebd7834b6
Deleted: sha256:d16e0fea83aa8c5c893878d8b7d1509b2449ce005e97bb359e44a8eceddd3ea9
Deleted: sha256:2235028ba88b6bdccbbdf56f3403c3d13731b253a227ec4e72674d34292598f3
Deleted: sha256:1578fe5946dbccf5ce0ed470766435a9136c5f5d48835eca47833536135800a3
Deleted: sha256:9e6c7628a8879594e2cd77dd73be3094f06f2ff11ee22942618e7fd916b8bf3b
Deleted: sha256:21545dba3c274ca7b7cb19ae6e3f4d5a96b9fb751295e4f4791d0165530f2dfe
Deleted: sha256:e1f001d2d22860ead8c04b64fd17428bc21d36260e248d4e101447986436482f
Deleted: sha256:043aa6e8e6e2cfecd2d91e16e00b80ad6159f9030d495eeceb694ef68735a05f
Deleted: sha256:11655cfc99cc816b9235e4db5770dc9e6f73254003f1d79502b18fa3cf425a08
Deleted: sha256:2b9692c22ea385e16af97ff3853a4c8930bc5bcd16db78532541120b219de69e
22: Pulling from heroku/builder
Digest: sha256:b3b0d2bb98243064670a4537ebc60bae0ebc669012bb6718285f1e3cf8768c44
Status: Image is up to date for heroku/builder:22
22-cnb: Pulling from heroku/heroku
Digest: sha256:726e37269cbdcd8d03499fbc6c9bf3d6fa7cf0b910ab2db0f3418ddc85407a60
Status: Image is up to date for heroku/heroku:22-cnb
===> ANALYZING
Image with name "my-image" not found
===> DETECTING
heroku/ruby 2.0.0
===> RESTORING
===> BUILDING

# Heroku Ruby Buildpack

- Ruby version `3.1.3` from `default`
  - Installing.......(4.671s)
- Bundler version `2.4.5` from `default`
  - Running `gem install bundler --version 2.4.5`....(1.328s)
- Bundle install
  - Running `BUNDLE_BIN="/layers/heroku_ruby/gems/bin" BUNDLE_CLEAN="1" BUNDLE_DEPLOYMENT="1" BUNDLE_GEMFILE="/workspace/Gemfile" BUNDLE_PATH="/layers/heroku_ruby/gems" BUNDLE_WITHOUT="development:test" bundle install`

      Fetching gem metadata from https://rubygems.org/..
      Fetching rake 13.0.6
      Installing rake 13.0.6
      Using bundler 2.4.5
      Fetching rack 2.2.3
      Fetching webrick 1.7.0
      Installing webrick 1.7.0
      Installing rack 2.2.3
      Bundle complete! 3 Gemfile dependencies, 4 gems now installed.
      Gems in the groups 'development' and 'test' were not installed.
      Bundled gems are installed into `/layers/heroku_ruby/gems`

  - Done (4.622s)
- Setting default processes(es)
  - Detecting gems
  - Running `bundle list`...(0.208s)
  - Detected rack app (`rack` gem and `config.ru` at root of application)
- Rake assets install
  - Cannot run rake tasks (no `Rakefile`)
  ! Help: Add `Rakefile` to your project to enable
- Done (finished in 10.927s)
```

Reply:
Pack uses `[=====>     ]` (but then clears them after they're done)

Yeah, I like this style, but `pack` has the advantage of knowing it's in a terminal that supports clears. Output edits just don't work well in dashboard, and we should probably avoid it.

Personally, I think `Running `bundle list`...(0.208s)` looks great, though I'd probably add a space after the last dot. (like `Running `bundle list`... (0.208s)`).

I think dim text has fairly limited support in terminals, but they should fall back to standard text without issue.